sled-agent/src/bootstrap/agent.rs

```diff
@@ -374,41 +365,61 @@ impl Agent {
         // the switch zone.
         info!(log, "Bootstrap Agent monitoring for hardware");
 
+        let hardware_monitor = Self::hardware_monitor(
+            &ba_log,
+            &config.link,
+            &sled_config,
+            global_zone_bootstrap_link_local_address,
+        )
+        .await?;
+
+        let storage_resources = hardware_monitor.storage().clone();
+
         let agent = Agent {
             log: ba_log,
             parent_log: log,
             ip,
             rss_access: Mutex::new(()),
             share: Mutex::new(None),
-            sled_state: Mutex::new(SledAgentState::Before(None)),
+            sled_state: Mutex::new(SledAgentState::Before(Some(
+                hardware_monitor,
+            ))),
+            storage_resources,
             config: config.clone(),
             sled_config,
             sp,
             ddmd_client,
             global_zone_bootstrap_link_local_address,
         };
 
-        let hardware_monitor = agent.start_hardware_monitor().await?;
-        *agent.sled_state.lock().await =
-            SledAgentState::Before(Some(hardware_monitor));
+        // Wait for at least the M.2 we booted from to show up.
+        //
+        // This gives the bootstrap agent a chance to read locally-stored
+        // configs if any exist.
+        loop {
+            match agent.storage_resources.boot_disk().await {
+                Some(disk) => {
+                    info!(agent.log, "Found boot disk M.2: {disk:?}");
+                    break;
+                }
+                None => {
+                    info!(agent.log, "Waiting for boot disk M.2...");
+                    tokio::time::sleep(core::time::Duration::from_millis(250))
+                        .await;
+                }
+            }
+        }
```

Comment:
This is new. Without this, it's possible that we simply don't catch our own bootdisk, due to the storage manager lagging behind.

I could possibly update this to less of a "retry-based" interface, but it seems to work.

Reply:
This makes sense to me. Seems fine to retry, since we need M.2s to boot.